s3torchconnector/src/s3torchconnector/s3reader/dcp_optimized.py

```diff
@@ -0,0 +1,588 @@
+            if r.start < items[-1].end:
+                if r.start < items[-1].start:
+                    raise ValueError(
+                        f"Unsorted ranges: {items[-1].start}-{items[-1].end} and {r.start}-{r.end}"
+                    )
+                else:
+                    raise ValueError(
+                        f"Overlapping ranges: {items[-1].start}-{items[-1].end} and {r.start}-{r.end}"
+                    )
+            # Coalesce or create new group
+            if r.start - items[-1].end <= max_gap_size:
+                items.append(r)
+            else:
+                group = RangeGroup(items[0].start, items[-1].end, items)
+                groups.append(group)
+                self._group_start_to_group[items[0].start] = group
+                items = [r]
+
+        final_group = RangeGroup(items[0].start, items[-1].end, items)
+        self._group_start_to_group[items[0].start] = final_group
+        groups.append(final_group)
+        return groups
+
+    def _find_item_for_position(self, pos: int) -> ItemRange:
+        """Find which item contains the given position with validations."""
+
+        if pos < self._current_item.start:
+            raise ValueError(
+                f"{FIND_ITEM_ERROR_PREFIX}Position {pos} before current range "
+                f"{self._current_item.start}-{self._current_item.end}"
+            )
+
+        # Return item if position still in current item
+        if pos < self._current_item.end:
+            return self._current_item
+
+        # Check next item
+        prev_item = self._current_item
+        try:
+            item = next(self._item_iter)
+
+            if pos < item.start:
+                raise ValueError(
+                    f"{FIND_ITEM_ERROR_PREFIX}Position {pos} in gap between ranges "
+                    f"{prev_item.start}-{prev_item.end} and {item.start}-{item.end}"
+                )
+            # Return item if position is in new item
+            if pos < item.end:
+                return item
+            else:
+                raise ValueError(
+                    f"{FIND_ITEM_ERROR_PREFIX}Position {pos} beyond next range "
+                    f"{item.start}-{item.end}"
+                )
+        except StopIteration:
+            raise ValueError(
+                f"{FIND_ITEM_ERROR_PREFIX}Position {pos} beyond last range "
+                f"{prev_item.start}-{prev_item.end}"
+            )
+
+    def _get_stream_for_item(self, item: ItemRange) -> GetObjectStream:
+        """Find which RangeGroup contains the given position."""
+
+        # If item is the first item of a new group, create new stream
+        if item.start in self._group_start_to_group:
+            group = self._group_start_to_group[item.start]
+            self._stream = self._get_stream(group.start, group.end)
+            self._stream_pos = group.start
+            self._leftover = None
+            return self._stream
```

Comment:
I feel there should be a subclass/dataclass just for handling the attributes on the stream